python/hidet/graph/ops/definitions/reduce/reduce.py

```diff
@@ -70,23 +68,158 @@ def reduce_fcompute(*reduce_indices):
+        from hidet.lang.cuda import blockIdx, threadIdx
+
+        row_major = DataLayout.row_major
+
+        warp_size = 32
+        block_size = warp_size
+        x, y = self.inputs[0], self.outputs[0]
+        xdtype = x.type.dtype
+        shape: List[int] = list(x.const_shape)
+        dims = self.dims
+        if self.keep_dim:
+            remain_shape = [v if i not in dims else 1 for i, v in enumerate(shape)]
         else:
-            # last dimension has not been reduced
-            return cuda_schedule_reduce_by_default(self)
+            remain_shape = [v for i, v in enumerate(shape) if i not in dims]
+        reduce_shape = [shape[i] for i in dims]
+        reduce_extent = hidet.utils.prod(reduce_shape)
+
+        remain_layout = spatial(*remain_shape)
+        layout = row_major(shape)
+
+        spatial_shape = []
+        repeat_shape = []
+        for i in range(len(shape)):
+            if i == len(shape) - 1:
+                spatial_shape.append(warp_size)
+                repeat_shape.append((shape[i] + warp_size - 1) // warp_size)  # num warps per row
+            elif i in dims:
+                spatial_shape.append(1)
+                repeat_shape.append(shape[i])
+            else:
+                spatial_shape.append(shape[i])
+                repeat_shape.append(1)
+        task_layout = repeat(*repeat_shape) * spatial(*spatial_shape)
+        grid_size = remain_layout.num_workers
+        accumulate_dtype = self.attrs['accumulate_dtype']
+        reduce_type = self.attrs['reduce_type']
+        ro = ReduceOperation.from_name(reduce_type)
+
+        with hidet.script_module() as module:
+
+            @hidet.script
+            def reduce_kernel(x: xdtype[x.const_shape], y: xdtype[y.const_shape]):
+                attrs.cuda.grid_dim = grid_size
+                attrs.cuda.block_dim = block_size
+                attrs.cuda.min_blocks = 1
+
+                rv = ro.initial_value(data_type(accumulate_dtype))
+                for indices in task_layout.on(threadIdx.x + blockIdx.x * block_size):
+                    if layout.within_bound(indices):
+                        k = x.read(indices, protected=False)
+                        rv = ro.combine(rv, cast(k, accumulate_dtype))
+                # Warp reduce by shuffle down
+                mask = active_mask()
+                rv = ro.combine(rv, shfl_down_sync(mask, rv, 16, 32))
+                rv = ro.combine(rv, shfl_down_sync(mask, rv, 8, 32))
+                rv = ro.combine(rv, shfl_down_sync(mask, rv, 4, 32))
+                rv = ro.combine(rv, shfl_down_sync(mask, rv, 2, 32))
+                rv = ro.combine(rv, shfl_down_sync(mask, rv, 1, 32))
+                rv = shfl_sync(mask, rv, 0, 32)
+                rv = ro.finalize(acc=rv, size=reduce_extent)
+
+                if threadIdx.x == 0:
+                    for indices in remain_layout.on(blockIdx.x):
+                        y.write(indices, cast(rv, xdtype), protected=False)
```

Comment:
```suggestion
                        y[indices] = cast(rv, xdtype)
```
I prefer above style of assignment, when we do not need protectation.